File: Algorithm/src/DFS/LeetCode98.java
package DFS;

public class LeetCode98 {

        class TreeNode{
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode(){}
        TreeNode(int val){this.val = val;}
        TreeNode(int val, TreeNode left, TreeNode right){
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    public boolean isValidBST(TreeNode root) {
        return DFS(root,null,null);
    }

    private boolean DFS(TreeNode root, Integer min,Integer max){
        if(root == null){
            return true;
        }
        if((max != null && root.val >= max) || (min != null && root.val <= min)){
            return false;
        }
        return DFS(root.left,min,root.val) && DFS(root.right,root.val,max);
    }
}

File: top-interview-questions-easy/tree/LeetCode98.go
package tree

import "math"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func isValidBST(root *TreeNode) bool {
	return validBST(root, math.MaxInt64, math.MaxInt64)
}

func validBST(root *TreeNode, min_val, max_val int64) bool {
	if root == nil {
		return true
	}
	if root.Val <= int(min_val) || root.Val >= int(max_val) {
		return false
	}
	return validBST(root.Left, min_val, int64(root.Val)) && validBST(root.Right, int64(root.Val), max_val)
}
